flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/asyncprocessing/tvf/common/AsyncStateWindowAggOperator.java

```diff
@@ -0,0 +1,325 @@
+    }
+
+    @Override
+    public void close() throws Exception {
+        super.close();
+        collector = null;
+        windowProcessor.close();
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        RowData inputRow = element.getValue();
+        RowData currentKey = (RowData) getCurrentKey();
+        windowProcessor
+                .processElement(currentKey, inputRow)
+                .thenAccept(
+                        isElementDropped -> {
+                            if (isElementDropped) {
+                                // markEvent will increase numLateRecordsDropped
+                                lateRecordsDroppedRate.markEvent();
+                            }
+                        });
+    }
+
+    @Override
+    public Watermark preProcessWatermark(Watermark mark) throws Exception {
+        if (mark.getTimestamp() > currentWatermark) {
+            // If this is a proctime window, progress should not be advanced by watermark, or it'll
+            // disturb timer-based processing
+            if (isEventTime) {
+                windowProcessor.advanceProgress(null, mark.getTimestamp());
+            }
+            return super.preProcessWatermark(mark);
+        } else {
+            return super.preProcessWatermark(new Watermark(currentWatermark));
+        }
+    }
+
+    @Override
+    public void onEventTime(InternalTimer<K, W> timer) throws Exception {
+        onTimer(timer);
+    }
+
+    @Override
+    public void onProcessingTime(InternalTimer<K, W> timer) throws Exception {
+        StateFuture<Void> advanceFuture = AsyncStateUtils.REUSABLE_VOID_STATE_FUTURE;
+        if (timer.getTimestamp() > lastTriggeredProcessingTime) {
+            // similar to the watermark advance,
+            // we need to notify WindowProcessor first to flush buffer into state
+            lastTriggeredProcessingTime = timer.getTimestamp();
+            advanceFuture =
+                    windowProcessor.advanceProgress(
+                            (RowData) getCurrentKey(), timer.getTimestamp());
+            // timers registered in advanceProgress() should always be smaller than current timer
+            // so, it should be safe to trigger current timer straightforwards.
+        }
+        advanceFuture.thenAccept(VOID -> onTimer(timer));
+    }
+
+    private void onTimer(InternalTimer<K, W> timer) throws Exception {
+        W window = timer.getNamespace();
+        windowProcessor
+                .fireWindow(timer.getTimestamp(), window)
+                .thenAccept(VOID -> windowProcessor.clearWindow(timer.getTimestamp(), window));
+        // we don't need to clear window timers,
+        // because there should only be one timer for each window now, which is current timer.
+    }
+
+    @Override
+    public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+        windowProcessor.prepareCheckpoint();
```

Comment:
It's better to perform `super.prepareSnapshotPreBarrier(long checkpointId)` before this.